### PR TITLE
Documentation deployment

### PR DESCRIPTION
**Pull request by alexschroeter**
With this branch, I want to get the automatic building and deployment to GitHub pages going.
We should have one site ( https://multiview-stitcher.github.io) for documentation and automatically deploy with GitHub Actions.

I would like to establish a pipeline where a feature branch can only be merged into dev. All changes that go to the main branch and get tagged with a release come from the dev branch.

### Documentation Features
- With `mike` we get versioned documentation. (This will be ['latest','dev', 'v1.1.1', ...])
- With `mkdocstrings` we get API documentation. (I have created an example docstring for the function register from the registration.py)

### Workflows
Therefore, I have created two workflows for the documentation.

For the dev branch:
Whenever there is a merge to dev the documentation with the version "dev" gets overwritten and the documentation is updated.

For the main branch:
Whenever there is a release tagged the documentation for that specific version is built and deployed.

### CODE REVIEW

File: .github/workflows/documentation_dev.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+name: Generate dev documentation
+
+on:
+  push:
+    branches:
+      - dev
+
+jobs:
+  api-docs:
```
Comment (m-albert) [inline .github/workflows/documentation_dev.yml:9]: Is there a reason this job is called api-docs when it pushes all docs (not only the API specific ones)?

File: src/multiview_stitcher/registration.py [hunk 2]
```diff
@@ -450,12 +450,35 @@ def get_node_params_from_reg_graph(g_reg):
...
-    transform_key=None,
-    new_transform_key=None,
-    registration_binning=None,
-):
+    msims: [multiscale_spatial_image],
+    reg_channel_index: int = None,
+    transform_key: unknown = None,
+    new_transform_key: unknown = None,
+    registration_binning: unknown = None,
+) -> [parametes]:
```
Comment (m-albert) [inline src/multiview_stitcher/registration.py:458]: Found a typo here :)

**Review by m-albert — COMMENTED** [review]
Thank you for this PR @alexschroeter.

> With this branch, I want to get the automatic building and deployment to GitHub pages going.  
We should have one site ( [https://multiview-stitcher.github.io](https://multiview-stitcher.github.io/)) for documentation and automatically deploy with GitHub Actions.

I think having available the docs for different releases and the development version is super useful. Your setup looks good to me (my experience is limited here though) and I added some review comments.

> I would like to establish a pipeline where a feature branch can only be merged into dev. All changes that go to the main branch and get tagged with a release come from the dev branch.

Do I understand you correctly that in this model all PRs would be merged into dev? If so, what would be the advantage of going over dev first? And at what point would they be merged into main? In my experience projects like `dask-image` and `scikit-image` would merge PRs directly into main, which would be ahead of the last release. Generally I'd think it makes sense to stick to what the neighboring community is doing, but I'm happy to have a discussion about this.

**Comment by alexschroeter** [comment]
So having a dev branch (and maybe also staging) branch seems to have gotten out of style with the possibilities of more elaborate CI/CD pipelines. 

I did a bit of reading on what other people do and we can certainly follow the pack. I think what you are describing is the _GitHub flow_ way of doing it.

I will adapt the documentation deployment to instead of having a _dev_ version from the dev branch to always have a _latest_ (maybe can come up with a better name because latest mostly refers to the latest version i.e. v1.3.4 is latest) version which will be the latest merge to main.

File: src/multiview_stitcher/registration.py (hunk 2 above)
Comment (alexschroeter) [reply src/multiview_stitcher/registration.py:458]: I will fix the typo, but this was just for me to have a test case for the documentation. I think the return type should also be more explanatory i.e. _transformation matrix 4x4_ or whatever it is.

File: .github/workflows/documentation_dev.yml (hunk 1 above)
Comment (alexschroeter) [reply .github/workflows/documentation_dev.yml:9]: I was rushing a bit to get the documentation working. I think I just copy and pasted from the previous version where I had multiple pipelines for the different parts of the documentation. **Will fix**.

**Comment by m-albert** [comment]
The model I was thinking of is indeed the [github flow](https://docs.github.com/en/get-started/quickstart/github-flow) as you mention (thanks for the keywords!). Great, let's follow the pack then.

> I will adapt the documentation deployment to instead of having a dev version from the dev branch to always have a latest (maybe can come up with a better name because latest mostly refers to the latest version i.e. v1.3.4 is latest) version which will be the latest merge to main.

Perfect. Regarding the name, I agree that _latest_ would mostly be understood to refer to the most recent release. Maybe one idea would be to use both the names _latest_ and _main_  to refer to the most recent release (there'd be some redundancy, but maybe that doesn't harm) and the repo's main branch which might be ahead of the last release?

File: src/multiview_stitcher/registration.py (hunk 2 above)
Comment (m-albert) [reply src/multiview_stitcher/registration.py:458]: Agreed.

**Review by m-albert — APPROVED** [review]
Looks great to me.

**Comment by m-albert** [comment]
Is the PR ready for merging?

**Comment by alexschroeter** [comment]
I think it should be ready. The actions stuff is a bit hard to test before hand, or I haven't found a way to do it.
But you can merge it if you want.

**Comment by m-albert** [comment]
Thanks. Yeah since the actions explicitly reference the main branch it's not clear how to best test them. There seems to be [this](https://github.com/actions/toolkit). 

But since the documentation deployment is anyways not working yet on main, let's merge and see :)

**Comment by m-albert** [comment]
Here's a little update after playing around a bit.

After merging, there was an [error](https://github.com/multiview-stitcher/multiview-stitcher/actions/runs/6534060199) in the github actions workflow. It's related to the permissions of the bot trying to push to the gh-pages branch of the repo.

So I played around a bit directly on the main branch by trying to change permissions and using a [github actions workflow](https://github.com/peaceiris/actions-gh-pages), but I'm not succeeding so far. Will be back at this tomorrow!

**Comment by alexschroeter** [comment]
I already saw that you were working on it and I sent you an E-Mail with what I believe to be the problem but you probably haven't seen it yet :D

**Comment by m-albert** [comment]
Haha exactly right, sorry I completely ignored emails in the last hour(s) somehow. Your suggestion seem to have worked! Now we're [here](https://github.com/multiview-stitcher/multiview-stitcher/actions/runs/6535419814). Rushing away for now though!

**Comment by alexschroeter** [comment]
Okay, I fixed that by allowing the gh-pages env to deploy from the gh-pages branch.
Now, you just have to revert the changes to the workflow, (maybe you also need to remove the gh-pages branch to clean up) and rerun the main documentation workflow to recreate the correct files.

**Comment by m-albert** [comment]
> Okay, I fixed that by allowing the gh-pages env to deploy from the gh-pages branch.

Cool, well spotted. The docs are online and I added them to the repo description: https://multiview-stitcher.github.io/multiview-stitcher/

> Now, you just have to revert the changes to the workflow, (maybe you also need to remove the gh-pages branch to clean up) and rerun the main documentation workflow to recreate the correct files.

Done, reverted my changes from yesterday. Doc deployment seems to work fine without deleting the gh-pages branch.

Thanks @alexschroeter for your work on this 🚀

**Comment by alexschroeter** [comment]
Yes, it's working but the gh-pages repo is a bit polluted now. You can see that there is almost all files of the main documentation in the root folder and there should only be like 3 files. All the documentation should just be in the main folder which is where index.html forwards to.

**Comment by m-albert** [comment]
You're right! Deleted the gh-pages branch and rerun the workflow. Now the dropdown menu showing the versions is also visible (which hadn't been before).

**Comment by m-albert** [comment]
Let's keep in mind to check after our first release that also that workflow is working.

**Pull request merged**